src/realm/util/sensitive_buffer.cpp

```diff
@@ -0,0 +1,227 @@
+    REALM_ASSERT_RELEASE_EX(ret != 0 && "VirtualFree()", GetLastError());
+
+    m_buffer = nullptr;
+}
+
+void SensitiveBufferBase::protect() const
+{
+    // MEMO even though we try to lock the page with the buffer, and that should prevent it to be swapped,
+    //      locking is not reliable and may fail under high demand in the system due to some opaque reason,
+    //      which we can't recover from, so use second layer to protect the buffer if it is swapped
+    //      note: look at attempt_to_lock where it's expected
+    BOOL ret = CryptProtectMemory(m_buffer, DWORD(m_size), CRYPTPROTECTMEMORY_SAME_PROCESS);
+    REALM_ASSERT_RELEASE_EX(ret == TRUE && "CryptProtectMemory()", GetLastError());
+}
+
+void SensitiveBufferBase::unprotect() const
+{
+    BOOL ret = CryptUnprotectMemory(m_buffer, DWORD(m_size), CRYPTPROTECTMEMORY_SAME_PROCESS);
+    REALM_ASSERT_RELEASE_EX(ret == TRUE && "CryptUnprotectMemory()", GetLastError());
+}
+
+#else
+SensitiveBufferBase::SensitiveBufferBase(size_t size)
+    : m_size(size)
+{
+    m_buffer = mmap(nullptr, m_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    REALM_ASSERT_RELEASE_EX(m_buffer != MAP_FAILED && "mmap()", errno);
+
+    int ret;
+    ret = mlock(m_buffer, m_size);
+#if REALM_LINUX
+    // mlock failing with ENOMEM under linux implies we're over RLIMIT_MEMLOCK,
+    // which happens inside a container. ignore for now.
+    int err = errno;
+    REALM_ASSERT_RELEASE_EX((ret == 0 || err == ENOMEM) && "mlock()", err);
+#else
+    REALM_ASSERT_RELEASE_EX(ret == 0 && "mlock()", errno);
+#endif
+
+
+#if defined(MADV_DONTDUMP)
+    // Linux kernel 3.4+ respects MADV_DONTDUMP, ignore return value on older
+    madvise(m_buffer, m_size, MADV_DONTDUMP);
+#endif
+}
+
+SensitiveBufferBase::~SensitiveBufferBase()
+{
+    if (!m_buffer)
+        return;
+
+    secure_erase(m_buffer, m_size);
+
+#if defined(MADV_DONTDUMP) && defined(MADV_DODUMP)
+    madvise(m_buffer, m_size, MADV_DODUMP);
+#endif
+
+    int ret;
+    ret = munlock(m_buffer, m_size);
+    REALM_ASSERT_RELEASE_EX(ret == 0 && "munlock()", errno);
+
+    ret = munmap(m_buffer, m_size);
+    REALM_ASSERT_RELEASE_EX(ret == 0 && "munmap()", errno);
+
+    m_buffer = nullptr;
+}
+
+void SensitiveBufferBase::protect() const {}
+
+void SensitiveBufferBase::unprotect() const {}
+#endif
+
+SensitiveBufferBase::SensitiveBufferBase(const SensitiveBufferBase& other)
+    : SensitiveBufferBase(other.m_size)
+{
+    std::memcpy(m_buffer, other.m_buffer, m_size);
+}
+
+SensitiveBufferBase::SensitiveBufferBase(SensitiveBufferBase&& other) noexcept
+    : m_size(other.m_size)
+{
+    std::swap(m_buffer, other.m_buffer);
+}
+
+void SensitiveBufferBase::secure_erase(void* buffer, size_t size)
+{
+#ifdef _WIN32
+    SecureZeroMemory(buffer, size);
+#elif defined(__STDC_LIB_EXT1__) || __APPLE__
+    memset_s(buffer, size, 0, size);
+#elif defined(__GNU_LIBRARY__) && __GLIBC__ >= 2 && __GLIBC_MINOR__ >= 25
+    explicit_bzero(buffer, size);
```

Comment:
This won't find musl's `explicit_bzero` - I think we're going to need to check for its existence at configure-time like we do with `readdir64`, for example.

Reply:
We actually had a build with Alpine to make sure we work with the musl C library on Jenkins. It's one of the things we'll need to bring back.

Reply:
ok, tried to build with Alpine and added support for this. Should work now.

Reply:
So we didn't even need the `_GNU_SOURCE` define? Awesome!